### DISCUSSION TITLE
[RFC] Usage of `freerdp_settings_* functions

### DISCUSSION THREAD

**Discussion by hardening** [FreeRDP/FreeRDP#9431] [General]
### General

I wanted to open the discussion about the usage of `freerdp_settings_get_<type>` and `freerdp_settings_set_<type>` in FreeRDP, as I've seen in some remarks in reviews that it's a subject that have not been discussed, and at least that there's no consensus around it.

### Public API
So my position is that making `rdpSettings` opaque for public API could be fine, and so at the end users of FreeRDP lib shall do `freerdp_settings_[get|set]` to interact with the settings. Anyway I also see some drawbacks to that, look forward to usage of these internally.

### Internal usage

On the other hand I'm totally opposed to generalizing the usage of these functions inside FreeRDP itself, but let's take a concrete example, so that it's not an abstract talk. We would have
```C
settings->NegotiateSecurityLayer = FALSE;
settings->VmConnectMode = TRUE;
settings->NlaSecurity = TRUE;
settings->AuthenticationLevel = 0;
```
that would become:
```C
if (!freerdp_settings_set_bool(settings, FreeRDP_NegotiateSecurityLayer, FALSE) ||
    !freerdp_settings_set_bool(settings, FreeRDP_VmConnectMode, TRUE) ||
    !freerdp_settings_set_bool(settings, FreeRDP_NlaSecurity, TRUE) ||
    !freerdp_settings_set_uint32(settings, FreeRDP_AuthenticationLevel, 0)) 
{
   /* error treatment */
}

```

I see major flaws with that:

* readability and error checks: a simple assignment becomes a function call with error checks. So potentially the initial code could be a function returning `void`, but now we have to return a `BOOL`, and force the callers to do checks if we want to have perfect error checking. I'm super attached to the looking of the code, so as much as possible no unneeded if statements, when reading the code that must be crystal clear, you must not wonder if  `freerdp_settings_set_bool` could legitimately fail (this is diverting attention);
* silent and error prone: if you do `settings->ServerHostname = TRUE`  the compiler will complain at build time, if you do `freerdp_settings_set_bool(settings, FreeRDP_ServerHostname, TRUE)` you'll discover the error at runtime (if you're lucky enough that this piece of code is commonly ran);
* performances at runtime and build time: we should not rely on the compiler guys to do some magics. I barely think that the compiler will generate `settings->NlaSecurity = TRUE` with `freerdp_settings_set_bool(settings, FreeRDP_NlaSecurity, TRUE)` even with all compiler tricks. I'd rather imagine in the best case one or more big jumptables (the settings ids for bools aren't contiguous) with tons of operations before actually doing the assignment. Also think how much build time computation the compiler would need to spend to figure out that `freerdp_settings_set_bool` could be simplified to a variable assignment (if that's even possible for the compiler to reach that conclusion);
* added value: and finally I can't see any added value to using that extra level of abstraction inside FreeRDP itself. We've used `freerdp_settings_set_string` for convenience in many places internally, in batch treatments, and to handle the freeing of the previous value, but honestly we should replace `freerdp_settings_set_string(settings, FreeRDP_ServerHostname, "localhost")` by `freerdp_update_string(&settings->ServerHostname, "localhost")` so the variable type can be statically checked. We don't need to hide how `rdpSettings` is structured to the rest of FreeRDP.

To conclude, you will have probably noticed that some of the arguments I have pointed for internal usage, also applies for public usage

**Comment by akallabeth** [comment]
So, here my POV:

> *    readability and error checks: a simple assignment becomes a function call with error checks. So potentially the initial code could be a function returning void, but now we have to return a BOOL, and force the callers to do checks if we want to have perfect > error checking. I'm super attached to the looking of the code, so as much as possible no unneeded if statements, when reading the > code that must be crystal clear, you must not wonder if freerdp_settings_set_bool could legitimately fail (this is diverting attention);

that is the price to pay for error checks at runtime. how else do you want to propagate some issue here?

> *   silent and error prone: if you do settings->ServerHostname = TRUE the compiler will complain at build time, if you do freerdp_settings_set_bool(settings, FreeRDP_ServerHostname, TRUE) you'll discover the error at runtime (if you're lucky enough that this piece of code is commonly ran);

point for improvement, yes. we could add that to the getter/setter generator to have a static assert for valid key values

> *    performances at runtime and build time: we should not rely on the compiler guys to do some magics. I barely think that the compiler will generate settings->NlaSecurity = TRUE with freerdp_settings_set_bool(settings, FreeRDP_NlaSecurity, TRUE) even with all compiler tricks. I'd rather imagine in the best case one or more big jumptables (the settings ids for bools aren't contiguous) with tons of operations before actually doing the assignment. Also think how much build time computation the compiler would need to spend to figure out that freerdp_settings_set_bool could be simplified to a variable assignment (if that's even possible for the compiler to reach that conclusion);

negligible, we´re talking settings here not some hot IO buffer swapping

> *    added value: and finally I can't see any added value to using that extra level of abstraction inside FreeRDP itself. We've used freerdp_settings_set_string for convenience in many places internally, in batch treatments, and to handle the freeing of the previous value, but honestly we should replace freerdp_settings_set_string(settings, FreeRDP_ServerHostname, "localhost") by freerdp_update_string(&settings->ServerHostname, "localhost") so the variable type can be statically checked. We don't need to hide how rdpSettings is structured to the rest of FreeRDP.

1. extensive runtime checks that are easy to add later on without touching every single piece of code working with settings
2. extensive compile time checks that are easily added/removed depending on necessity

**Comment by hardening** [comment]
> that is the price to pay for error checks at runtime. how else do you want to propagate some issue here?

That's why I gave an example: in that case (and in many others) in the original code, there's no possible error, readability is optimal. With the new version, readability is decreased a lot, when debugging you will always wonder if that set function could fail. There's tons of conditions and checks added, that forces to add error checking in the function itself and but also in callers even if there's no possible error. Now even functions that can't fail would seem like they could. It adds an enormous complexity everywhere in the code.

> point for improvement, yes. we could add that to the getter/setter generator to have a static assert for valid key values

Static assert are a C++ things and have nothing magic. The design can not rely on "the compiler will manage with that" (which it doesn't BTW). I'd be for refusing these changes until we've seen with the compiler output what the compiler can effectively check, optimize and measure the loss of performance (or the gain, who knows ?).
One thing that could already improve the public API would be that bool key ids, string key ids would be specific enums, so that setter/getter would take one of this enum as argument. So:
```C
typedef enum {
   FreeRDP_ServerName = 1001,
 ...
} freerdp_setting_string_t;

BOOL freerdp_settings_set_string(rdpSettings *settings, freerdp_setting_string_t id, const char* value);
```

> negligible, we´re talking settings here not some hot IO buffer swapping

Well sure, this is not an issue as long as you will not call too much these functions that will set/get these settings. But once all FreeRDP is treated how to be sure that you're not in these codepath ?

>
>1. extensive runtime checks that are easy to add later on without touching every single piece of code working with settings
>2 .extensive compile time checks that are easily added/removed depending on necessity
>

1. Well I must admit I can't see the real use case for "runtime checks on settings" ? What's the case where we need to monitor all the settings all the time ? And for that, every access to settings must be changed, is it worth having the whole FreeRDP codebase harder to read, kinda look bloated for such a limited usage ? Again look at my example, who wants the code to go from clear assignments, to complicated function calls ?
2. and for build time checks, in its current form using get and set function remove some checks. Without using getter/setter, having some more specific types for some of our settings, would already allow the compiler to do these build time checks. And that's probably going on our favorite dispute topic: where to place checks/asserts, where are the safe and tested area where you can trust the caller and avoid having assert everywhere. With the same thing that comes all the time: instead of adding asserts all over the place and sometime have the feeling of the code doing self-protection, let's increase the test coverage and build some safe arena where these asserts can be limited.

**Comment by akallabeth** [comment]
> > that is the price to pay for error checks at runtime. how else do you want to propagate some issue here?
> 
> That's why I gave an example: in that case (and in many others) in the original code, there's no possible error, readability is optimal. With the new version, readability is decreased a lot, when debugging you will always wonder if that set function could fail. There's tons of conditions and checks added, that forces to add error checking in the function itself and but also in callers even if there's no possible error. Now even functions that can't fail would seem like they could. It adds an enormous complexity everywhere in the code.
>
wrong. you can not do inter settings checks that way. that is the idea behind the getter/setter, make it opaque and run consistency checks in the background, depending on what you require (might not be used in production, but already saved a lot of time finding bugs where unknown messages override a value)
 
> > point for improvement, yes. we could add that to the getter/setter generator to have a static assert for valid key values
> 
> Static assert are a C++ things and have nothing magic. The design can not rely on "the compiler will manage with that" (which it doesn't BTW). I'd be for refusing these changes until we've seen with the compiler output what the compiler can effectively check, optimize and measure the loss of performance (or the gain, who knows ?). One thing that could already improve the public API would be that bool key ids, string key ids would be specific enums, so that setter/getter would take one of this enum as argument. So:
> 
yes, the language feature `static assert`, but you can do a simple `#define foo(x)  #if x != bar error out` or similar. also value range checks can be done this way (think the `Port` value for a simple example that uses a type larger than the allowed values)

> ```c
> typedef enum {
>    FreeRDP_ServerName = 1001,
>  ...
> } freerdp_setting_string_t;
> 
> BOOL freerdp_settings_set_string(rdpSettings *settings, freerdp_setting_string_t id, const char* value);
> ```
> 
> > negligible, we´re talking settings here not some hot IO buffer swapping
> 
> Well sure, this is not an issue as long as you will not call too much these functions that will set/get these settings. But once all FreeRDP is treated how to be sure that you're not in these codepath ?

because these functions are not expensive however you look at it.
even if they would be, in the worst case you can always do a `static INLINE` if one of these is really affecting speed.

> 
> > 1. extensive runtime checks that are easy to add later on without touching every single piece of code working with settings
> >    2 .extensive compile time checks that are easily added/removed depending on necessity
> 
>     1. Well I must admit I can't see the real use case for "runtime checks on settings" ? What's the case where we need to monitor all the settings all the time ? And for that, every access to settings must be changed, is it worth having the whole FreeRDP codebase harder to read, kinda look bloated for such a limited usage ? Again look at my example, who wants the code to go from clear assignments, to complicated function calls ?
> 
1. who is talking about all the settings all the time?
2. like value range checks (please do not bring a `BOOL` as example here)
3. inter dependency checks (e.g. you have setting a that requires b, so check they are both valid - here you can have the `BOOL` values too and it would be great to have it encapsulated)
4. simpler debugging (ever tried to find out who is setting something when? just add a print in the `getter` or `setter`)

>     2. and for build time checks, in its current form using get and set function remove some checks. Without using getter/setter, having some more specific types for some of our settings, would already allow the compiler to do these build time checks. And that's probably going on our favorite dispute topic: where to place checks/asserts, where are the safe and tested area where you can trust the caller and avoid having assert everywhere. With the same thing that comes all the time: instead of adding asserts all over the place and sometime have the feeling of the code doing self-protection, let's increase the test coverage and build some safe arena where these asserts can be limited.

where do you see a contradiction here?
1. the asserts are there so that if you change something unrelated you get a nice log message with stack trace if something essential fails. 
2. test coverage :+1: - already using the two to find issues all the time (hey, the `proxy` is quite a nice test case for that)
3. not `self protecting` - protecting from yourself if you missed some effect of a change (similar to how `ASAN` did eliminate the once memory leaking monster freerdp was)

**Comment by hardening** [comment]
### Settings dependencies

What's the percentage of settings that have dependencies ? A few of them come to my mind, but not that much. Let's do an inventory, is it a massive, or is it just a few cases ? Are these dependencies, things that always need to be handle at any time during the client/peer's life (that would kinda justify universal setters/getters), or dealing with the dependencies at some checkpoints is enough (just like with we do with parameters sanitizing and populating settings at the start of the connection) ?

If we really want to take some attentions to these settings dependencies, I'm almost sure that universal settings getter/setter inside FreeRDP is not the solution. Let's take an example of these deps:
* when we enable egfx, then we need drdynvc too
* when you set the network type it also "forces" a few other settings

So as a suggestion it's like these beasts might have a setter that will have side effects on other settings, but the getter can be trivial and access directly, so `freerdp_settings_set_channel_enabled(settings, channel_egfx, TRUE)` (returning `void`, not a `BOOL` that needs to be tested even if it will never fail) for a set op, and `settings->SupportGraphicsPipeline` is used directly.

### settings with restricted value set

Well from my PoV this is a public API concern only, inside FreeRDP we can ensure that we put a correct value. And the good news, is that most of the time we fill these settings with constants. So let's type more strongly some settings: network type shall be an enum, just like DesktopOrientation.

### instrumentation of settings

On the static analysis side, any modern IDE (even VSCode has it !) have a "find references" features that will list you instantly the functions that use a given setting. BTW the introduction of freerdp_settings setters/getters in the code introduced the burden that I have to do 2 researches: one on the setting itself and another on the associated key.

Shall we increase drastically the code complexity and readability to unable perfect settings instrumentation ? Shall we make our daily life miserable as soon as we interact with settings, for a very limited use case: my opinion is no, it's not worth it (never had to instrument settings to see which part of freerdp was getting or setting it).
And anyway if we really want to have settings instrumentation inside FreeRDP, it's not suitable to use the public API for this. For the performances, for readability, and in the current state we'd loose the static check done by the compiler.

And BTW in recent PRs when you systematically replace some trivial access to some uint32 or bool settings, you take the risk to use the wrong key and use the default value. And also some of these are in the "hot path" like in sending packets when reading various PDU limits and so you change a 1-2 cycles reader to a big function call with an enormous switch. With non contiguous setting id value, the compiler can't do any magic and it ends in the best case in discontinous jump tables, and in the worst case by testing every value. Some tries to optimize things in assembly in encode functions to save a few cycles, and we trivially loose 100 of them for settings instrumentation that is (almost) never used...

To conclude, let's do a concrete inventory of the needs with settings regarding dependencies, and value checks. Generalizing getters and setters inside FreeRDP with the public API doesn't seem like the good solution to address these issues.

**Comment by hardening** [comment]
Just to give an example of what the compiler generates for `freerdp_settings_get_bool` (There's no compiler magic (TM)), reversed in ghidra:

```C
undefined4 freerdp_settings_get_bool(long param_1,ulong param_2)

{
  int iVar1;
  undefined8 uVar2;
  undefined8 uVar3;
  
  if (param_1 == 0) {
    winpr_int_assert("settings","/home/david/dev/git/FreeRDP/libfreerdp/common/settings_getters.c",
                     "freerdp_settings_get_bool",0x38);
  }
  if (param_2 < 0x1447) {
    if (param_2 < 0x1440) {
      if (param_2 == 0x13c3) {
        return *(undefined4 *)(param_1 + 0x9e18);
      }
      if (param_2 < 0x13c4) {
        if (param_2 == 0x12c0) {
          return *(undefined4 *)(param_1 + 0x9600);
        }
        if (param_2 < 0x12c1) {
          if (param_2 == 0x1242) {
            return *(undefined4 *)(param_1 + 0x9210);
          }
          if (param_2 < 0x1243) {
            if (param_2 == 0x1241) {
              return *(undefined4 *)(param_1 + 0x9208);
            }
            if (param_2 < 0x1242) {
              if (param_2 == 0x1240) {
                return *(undefined4 *)(param_1 + 0x9200);
              }
              if (param_2 < 0x1241) {
                if (param_2 == 0x11c0) {
                  return *(undefined4 *)(param_1 + 0x8e00);
                }
                if (param_2 < 0x11c1) {
                  if (param_2 == 0x1141) {
                    return *(undefined4 *)(param_1 + 0x8a08);
                  }
                  if (param_2 < 0x1142) {
                    if (param_2 == 0x1140) {
                      return *(undefined4 *)(param_1 + 0x8a00);
                    }
                    if (param_2 < 0x1141) {
                      if (param_2 == 0x10c1) {
                        return *(undefined4 *)(param_1 + 0x8608);
                      }
                      if (param_2 < 0x10c2) {
                        if (param_2 == 0x10c0) {
                          return *(undefined4 *)(param_1 + 0x8600);
                        }
                        if (param_2 < 0x10c1) {
                          if (param_2 == 0x1044) {
                            return *(undefined4 *)(param_1 + 0x8220);
                          }
                          if (param_2 < 0x1045) {
                            if (param_2 == 0x1040) {
                              return *(undefined4 *)(param_1 + 0x8200);
                            }
                            if (param_2 < 0x1041) {
                              if (param_2 == 0xfc1) {
                                return *(undefined4 *)(param_1 + 0x7e08);
                              }
                              if (param_2 < 0xfc2) {
                                if (param_2 == 0xfc0) {
                                  return *(undefined4 *)(param_1 + 0x7e00);
                                }
                                if (param_2 < 0xfc1) {
                                  if (param_2 == 0xf80) {
                                    return *(undefined4 *)(param_1 + 0x7c00);
                                  }
                                  if ((param_2 < 0xf81) && (param_2 < 0xf0a)) {
                                    if (param_2 < 0xec0) {
                                      if (param_2 == 0xe85) {
                                        return *(undefined4 *)(param_1 + 0x7428);
                                      }
                                      if (param_2 < 0xe86) {
                                        if (param_2 == 0xe84) {
                                          return *(undefined4 *)(param_1 + 0x7420);
                                        }
                                        if (param_2 < 0xe85) {
                                          if (param_2 == 0xe80) {
                                            return *(undefined4 *)(param_1 + 0x7400);
                                          }
                                          if (param_2 < 0xe81) {
                                            if (param_2 == 0xe44) {
                                              return *(undefined4 *)(param_1 + 0x7220);
                                            }
                                            if (param_2 < 0xe45) {
                                              if (param_2 == 0xe41) {
                                                return *(undefined4 *)(param_1 + 0x7208);
                                              }
                                              if (param_2 < 0xe42) {
                                                if (param_2 == 0xe40) {
                                                  return *(undefined4 *)(param_1 + 0x7200);
                                                }
                                                if (param_2 < 0xe41) {
                                                  if (param_2 == 0xdc2) {
                                                    return *(undefined4 *)(param_1 + 0x6e10);
                                                  }

...
```

Reversed the function is 622 lines all of that kind (anyway it catches contiguous values), so the the 1 to 100 cycles ratio is probably underestimated (not even talking of how it kills the caches for the prediction branch unit)

**Comment by akallabeth** [comment]
@hardening not really an argument, as pointed out if we have something that is too slow  we can already make these functions inline (and have them optimized out for each lookup/set and still at one place to change/add logging/...). but show me the profiler that measures such a function using even 1% of processing power

* what we can not easily do with the current exposed struct is refactor the settings as they are still a mess of runtime configuration, initial settings and such.
* what we can not do is have an easy way to debug settings state changes as we need to change every line of code where a setting is touched
* we can not change the internals of the settings to be something doing some more error checking

so your only valid point is a more verbose syntax up until now and as you pointed out there are IDE that can autocomplete that for you already ;)

**Comment by hardening** [comment]
### Settings instrumentation
Clearly the public settings API is not the correct way to achieve this goal. It's slow, error prone and when used internally makes the code ugly. And I'd add that this is so a niche subject / need, that it's you who may prove that the gain are worth it (not me who should prove the loss). 

If we ever wants settings instrumentation, let's develop an API for internal-use-only that can switch at build time between the simple assignment and instrumented one. Something like:

```C
#ifdef WITH_SETTINGS_INSTRUMENTATION

inline BOOL checkBoolValFunc(BOOL v) {
    WINPR_UNUSED(v);
    return TRUE;
}

#define SETTINGS_GET_BOOL(settings, key) (checkBoolValFunc(settings->key) && freerdp_settings_get_bool(settings, FreeRDP_##key))
#define SETTINGS_SET_BOOL(settings, key, v) \
do { \
    (void)checkBoolValFunc(settings->key);
    freerdp_settings_set_bool(settings, FreeRDP_##key, v);
} while(0) 

#else

#define SETTINGS_GET_BOOL(settings, key) (settings->key)
#define SETTINGS_SET_BOOL(settings, key, v) settings->key = v

#endif
```

And in the code: `if (SETTINGS_GET_BOOL(settings, NlaSecurity)) { ....`

This way we have milk and butter: even if you're the only user of this, you can instrument settings changes by enabling `WITH_SETTINGS_INSTRUMENTATION` at build time, put breakpoints and printf in `freerdp_settings_set_bool`. With the call to `checkBoolValFunc()` you also have the type check at build time and for free at runtime (the compiler will trivially expand that to void). And the vast majority of people that don't need that won't pay the price. It will make the code a little heavier in terms of readability but at least we don't have the implied checks when setting a value.


### settings refactoring
If we come to make settings opaque for FreeRDP clients (public API only not internally), here's some thoughts:

* I also have the feeling that this part of FreeRDP is not very clean and that settings mix state and settings. Anyway at first look it seems like a big subject that may lead to discussions, proposals and PoCs and for sure the public settings API is not the solution to that;
* if settings are made opaque to the library client then we can deeply change the rdpSettings struct and if we have enums for the ids of string, bool and int settings, so that in the settings, bools, strings and int are packed in arrays of all bool settings, then the value can be retrieved by `settings->bools[id]`:
```C
typedef enum {
    FreeRDP_NlaSecurity = 0,
   ...
   FreeRDP_settings_n_bools
} freerdp_setting_id_bool;

struct rdp_settings {
    BOOL bools[FreeRDP_settings_n_bools];
    BOOL ints[FreeRDP_settings_n_ints];
    ...
};

BOOL freerdp_settings_get_bool(rdpSettings *settings, freerdp_setting_id_bool id) {
     if (id >= FreeRDP_settings_n_bools) {
        /* complain */
    }
    return settings->bools[id];
}
```
This way we're not relying on compiler capabilities to optimize a switch. That will also reduce (if not eliminate) the code generation part for the settings. Anyway that shall be discussed as it will make debugging like hell (or we may have this rdpSettings definition for release and a another one for debug).
* about settings with range values, well let do an inventory.

**Reply by akallabeth** [reply]
sounds like premature optimization to me.
also, replacing functions with macros and calling the former error prone is kind of contradicting ;)

anyway, all your suggestions require that there is an API (getter/setter) which is in place to allow changing the internals (which is the whole point of the procedure after all)

and a last point: I already mentioned you can already (with a minor change) just declare the functions `static inline`, include the getters/setters internally directly (not the public header) and there will be no difference to a macro, well, except for proper type checks.
 (just don´t see the point to spend too much time optimizing something that is not performance critical at all

**Reply by akallabeth** [reply]
as for the refactoring of the struct, for that we´d need to have the whole thing opaque first ;)
not really like to have to change all externals because we expose these internals in a public API

**Reply by hardening** [reply]
> also, replacing functions with macros and calling the former error prone is kind of contradicting ;)

Well, macros aren't error prone by nature. Error prone was because you discover a `freerdp_get_bool(settings, FreeRDP_ServerName)` only at runtime, in my proposal with macros `SETTINGS_GET_BOOL(settings,ServerName)` will not compile, you catch the error at build time.

**Reply by akallabeth** [reply]
@hardening that is a good poing.

currently the `FreeRDP_ServerName` et al are just defines.
We can easily change these to a specific `enum` and have that compile time check too.
so, maybe a good idea to review the API and change the `size_t` for the keys to a `typedef enum`

so, like :
```
typedef enum {
FreeRDP_ServerName = 23,
...
} FreeRDP_String_Settings_Key;
```
for each settings type
